# alignfix (CSE 185 Project)

Version: 0.1

Alignfix is a tool for aligning short Illumina reads to a genome based on a seed and extend method.
Alignifx uses fast pattern matching to rapidly find seeds in a genome and then proceeds to align the reads 
in the genome located around the seed. Alignfix consists of two key algorithmic ideas. For generating seeds,
Alignfix uses fast pattern matching by way of preprocessing the genome into suffix arrays. With a seed, alignfix creates
two different alignment on either side of the seed. Alignfix uses affine gap penalties for the alignment scoring
function.

## Install instructions
Installation requires the NumPy and Pyfaidx library. You must have Python 3.7 or later versions to install NumPy and Python 3.6 or later versions to install Pyfaidx.
You can install these libraries using the following command:
```
pip install numpy
pip install pyfaidx
```
Navigate to the directory in which you would like to download this tool and use the following command:
```
git clone https://github.com/msampath25/alignfix
```
Change into that directory and install the tool so that it can be used from the command line. You can do this with the following command:
```
cd alignfix
```
Once required libraries are installed, you can install ```alignfix``` with the following command: 
```
python setup.py install
```
Note: if you do not have root access, you can run the commands above with additional options to install locally:
```
pip install --user pyfaidx
python setup.py install --user
```
Alernatively, you can install ```alignfix``` into the home directory:
```
python setup.py install --prefix=$HOME
```
If the install was successful, typing ```python alignfix --help``` should show a useful message. 

## Mandatory Command Line Arguments
- `--genome`, `-g`: Fasta format of genome (this should be a single genome)
- `--query`, `-q`: Fasta format of queries
- `--output`, `-o`: File path for alignment output

## Usage:
The basic usage of ```alignfix``` is:
```
python alignfix --genome genome.fasta --query queries.fasta --output output.txt
```
To run ```alignfix``` on a small test example (using the files in this repo):
```
python alignfix --genome test_datasets/test1.fasta --query test_datasets/small_queries_test1.fasta --output test_output.txt
```
Optional arugments:

- `--bench`, `-b`: this will output information regarding benchmark stats


## Viral Example

We used this viral genome. You can find the data here: [Reads](https://www.ebi.ac.uk/ena/browser/view/PRJEB37886), [Genome](https://www.ncbi.nlm.nih.gov/nuccore/NC_045512.2)

This was the command we used to run the test dataset: 

```
python alignfix -g real_world_dataset/real_genome.fasta -q real_world_dataset/real_reads.fasta -o real_world_dataset/real_dataset_output.txt -b real_world_dataset/real_benchmark.txt
```

## File format
The output file format is a text file. Each sequence output in the text file should include the following in order: 
header (starts with ">"), two aligned sequences (database on top and query on the bottom), score, start position, and end position. Note that the start and end
position give the 0-based index of the database or genome sequence.

## Sources

- Altschul, S.F., Gish, W., Miller, W., Myers, E.W., Lipman, D.J. (1990) “Basic local alignment search tool.” J. Mol. Biol. 215:403-410
- Heng Li, Richard Durbin, Fast and accurate short read alignment with Burrows–Wheeler transform, Bioinformatics, Volume 25, Issue 14, July 2009, Pages 1754–1760, https://doi.org/10.1093/bioinformatics/btp324
- European Bioinformatics Institute. "Project: PRJEB37886." European Molecular Biology Laboratory, https://www.ebi.ac.uk/ena/browser/view/PRJEB37886.
- Wu, F, et al. “Severe Acute Respiratory Syndrome Coronavirus 2 Isolate Wuhan-Hu-1, Co - Nucleotide - NCBI.” National Center for Biotechnology Information, U.S. National Library of Medicine, 17 Jan. 2020, www.ncbi.nlm.nih.gov/nuccore/NC_045512.2. 

## Contributors
This repository was generated by Jason Chiu, Manish Sampath, and Siddharth Kaipa, with inspiration from [BLAST](https://blast.ncbi.nlm.nih.gov/Blast.cgi), [stack overflow](https://stackoverflow.com/questions/56534678/how-to-create-a-cli-in-python-that-can-be-installed-with-pip/66010978#66010978), and the CSE 185 Example projects .

Please submit a pull request with any other corrections or suggestions.

